File: Combinations/Combinations.py


class Solution:
    def dfs(self,pos,k,n,start,ans,vec):
        if pos==k:
            temp=vec[:]
            ans.append(temp)
            return 
        for i in range(start,n+1):
            vec[pos]=i
            self.dfs(pos+1,k,n,i+1,ans,vec)
    # @return a list of lists of integers
    def combine(self, n, k):
        ans=[]
        vec=[0 for i in range(k)]
        self.dfs(0,k,n,1,ans,vec)
        return ans
        
 

File: solutions/Combinations/Combinations.cpp

class Solution {
public:
    void dfs(vector<vector<int> > &ans,vector<int> vec,int n,int k,int pos,int start)
    {
        if(pos==k)
        {
            ans.push_back(vec);
            return ;
        }
        for(int i=start;i<=n;i++)
        {
            vec[pos]=i;
            dfs(ans,vec,n,k,pos+1,i+1);
        }
    }
    vector<vector<int> > combine(int n, int k) {
        vector<vector<int> >ans;
        vector<int>vec(k);
        dfs(ans,vec,n,k,0,1);
        return ans;
    }
};


 